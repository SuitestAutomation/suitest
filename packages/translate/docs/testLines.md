# Test Lines

Translating test line definitions into human readable language.

## Prerequisites

In order to translate a line, you'll need the following data:

* Line definition, e.g. loaded from feed `/results/{resultId}`.
* Application configuration, e.g. from feed `/test-pack-runs/{testPackRunId}`, see `effectiveAppConfig` property.
    This information is needed because any configuration variables used in test need to be translated.
* Optional, a list of elements used during test execution, so that actual element names could be displayed instead of
    just ids. This information will be available from the feed `/results/{resultId}` shortly.
* Optional, a list of snippets used during test execution, so that actual snippet names could be displayed instead of
    just ids.

Reference the feeds can be found in [our documentation].

## Translating test line definition to a plain text

For example:
```javascript
const {testLineToPlainText} = require('@suitest/translate');

const translateLine = testLineToPlainText({type: 'sleep', timeout: 2000}, appConfiguration);

console.log(translateLine); // Sleep for 2s
```

## Translating test line definition to a formatted text

```javascript
const {testLineToFormattedText} = require('@suitest/translate');

const translateLine = testLineToFormattedText({
    type: 'assert',
    then: 'success',
    condition: {
        subject: {type: 'application'},
        type: 'exited'
    },
}, appConfiguration);

console.log(translatedLine); // Assert: Application has exited
```

## Translating test line definition to a HTML

```javascript
const {testLineToHtml} = require('@suitest/translate');

const translateLine = testLineToHtml({
    type: 'wait',
    then: 'success',
    timeout: 2000,
    condition: {
        subject: {type: 'cookie', val: 'my-cookie'},
        type: '=',
        val: 'test'
    },
}, appConfiguration);

console.log(translatedLine);
```

Output (in HTML):
```html
<figure class="suitest-test-line">
	<figcaption>Assert: Cookie</figcaption>
	<figure class="suitest-test-line__dictionary">
		<dl>
			<dt>Timeout</dt>
			<dd><span class="suitest-test-line__text--bold">2s</span></dd>
		</dl>
	</figure>
	<figure class="suitest-test-line__table">
		<table>
			<tr>
				<td><span class="suitest-test-line__text--bold">my-cookie</span> cookie</td>
				<td>=</td>
				<td><span class="suitest-test-line__text--bold">test</span></td>
			</tr>
		</table>
	</figure>
</figure>
```

In order to apply basic styling, make sure to include `@suitest/translate/suitest-test-line.css` file from
the NPM package into your HTML page.

## Other formats

We are using our own flavour of [unist] AST to translate test lines into multiple formats. Thus, you can implement
your own renderers if provided by default do not fit your needs.

```javascript
const {testLineToAst} = require('@suitest/translate');

const suitestAst = testLineToAst({type: 'sleep', timeout: 2000});

console.log(suitestAst);
```

Would produce:
```json5
{
  type: 'test-line',
  title: [
    {type: 'text', value: 'Sleep for '},
    {type: 'bold', value: '2'},
    {type: 'text', value: 's'},
  ],
}
```

[our documentation]: https://suite.st/docs/suitest-network-api/api-reference/
[unist]: https://github.com/syntax-tree/unist
