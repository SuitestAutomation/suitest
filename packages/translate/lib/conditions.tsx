/// <reference path="../types/intrinsicElements.d.ts" />
/// <reference path="../types/unistTestLine.d.ts" />
import {jsx} from './jsxFactory';
import {
	Condition,
	CookieCondition,
	CurrentLocationCondition,
	ElementCondition,
	JavaScriptExpressionCondition,
	NetworkRequestCondition,
	Comparator,
	NetworkRequestInfo,
	AppConfiguration,
	Elements, ElementSubject, CustomElementSubject,
} from '@suitest/types';
import {formatVariables, replaceVariables} from './utils';
import {translateComparator} from './comparators';
import {ElementPropertiesCondition, PSVideoHaNoErrorCondition, PSVideoSubject} from '@suitest/types/lib';

const translateApplicationExitedCondition = (): ConditionNode =>
	<condition title="Application has exited"/> as ConditionNode;

const translateCurrentLocationCondition = (
	condition: CurrentLocationCondition,
	appConfig: AppConfiguration
): ConditionNode =>
	<condition title="Current location">
		<table>
			<row>
				<cell>Current location</cell>
				<cell>{translateComparator(condition.type)}</cell>
				<cell>{formatVariables(condition.val, appConfig.configVariables)}</cell>
			</row>
		</table>
	</condition> as ConditionNode;

const translateCookieCondition = (condition: CookieCondition, appConfig: AppConfiguration): ConditionNode =>
	<condition title="Cookie">
		<table>
			<row>
				<cell>{formatVariables(condition.subject.val, appConfig.configVariables)} cookie</cell>
				<cell>{translateComparator(condition.type)}</cell>
				<cell>{formatVariables(condition.val, appConfig.configVariables)}</cell>
			</row>
		</table>
	</condition> as ConditionNode;

// TODO - cover with unit tests
const translateElementProperty = (property: string): string => {
	switch (property) {
		// Special cases, when converting from cameCase is not enough
		case 'videoUrl':
			return 'video URL';
		case 'url':
			return 'URL';
		case 'color':
			return 'text color';
		case 'zIndex':
			return 'z-index';
		case 'automationId':
			return 'automation ID';
		case 'scaleX':
			return 'scale X';
		case 'scaleY':
			return 'scale Y';
		case 'translationX':
			return 'translation X';
		case 'translationY':
			return 'translation Y';
		case 'pivotX':
			return 'pivot X';
		case 'pivotY':
			return 'pivot Y';
		case 'tagInt':
			return 'tag';
		case 'fontURI':
			return 'font URI';
		case 'proposalURL':
			return 'proposal URL';
		default:
			// Split camel case to words
			return property.replace(/([A-Z+])/g, (_, match) => ' ' + match.toLowerCase());
	}
};

const assertUnknownElementCondition = (condition: never): never => {
	throw new Error(`Unknown element condition type: ${JSON.stringify(condition)}`);
};

const translateElementName = (subject: ElementSubject | PSVideoSubject, elements?: Elements): Node | Node[] => {
	if (subject.type === 'video' || (subject as CustomElementSubject).val?.video) {
		return <bold>video</bold>;
	}

	if (subject.type === 'psVideo') {
		return <bold>PlayStation 4 video</bold>;
	}

	if ('elementId' in subject) {
		// Element defined by it's ID
		if (elements && elements[subject.elementId]) {
			return <bold>{elements[subject.elementId].name}</bold>;
		}

		if (subject.name) {
			// Deprecated TODO remove after BE feeds updated
			return <bold>{subject.name}</bold>;
		}

		if (subject.nameHint) {
			// In case element was removed and we no longer have it on record
			return <bold>{subject.nameHint}</bold>;
		}

		return <fragment>
			Unknown element (<code>{subject.elementId.slice(0, 4) + '...' + subject.elementId.slice(-4)}</code>)
		</fragment>;
	}

	if ('apiId' in subject) {
		// Element defined by it's API ID
		return <bold>{subject.apiId}</bold>;
	}

	// Otherwise it's a custom element defined by it's selector
	const {ifMultipleFoundReturn, ...selector} = subject.val;
	const selectorKeys = Object.keys(selector);

	if (selectorKeys.length === 1) {
		// A common case when there is a single selector, e.g. css or xpath
		// Casting to any because TS throws an unwarranted error otherwise
		return <code>{(selector as any)[selectorKeys[0]]}</code>;
	}

	return <code>{JSON.stringify({...selector, index: ifMultipleFoundReturn ?? 1})}</code>;
};

const translateElementCondition = (
	condition: ElementCondition,
	appConfig: AppConfiguration,
	elements?: Elements
): ConditionNode => {
	const elementName = translateElementName(condition.subject, elements);

	switch (condition.type) {
		case 'exists':
			return <condition title={<fragment>{elementName} exists</fragment>}/> as ConditionNode;
		case '!exists':
			return <condition title={<fragment>{elementName} does not exist</fragment>}/> as ConditionNode;
		case 'matches':
			const codeReplacedVars = replaceVariables(condition.val, appConfig.configVariables);

			return <condition title={<fragment>{elementName} matches JavaScript</fragment>}>
				<code-block label="JavaScript matcher">{codeReplacedVars}</code-block>
				{codeReplacedVars !== condition.val
					? <code-block label="With variables">{condition.val}</code-block>
					: undefined
				}
			</condition> as ConditionNode;
		case 'visible':
			return <condition title={<fragment>{elementName} is visible</fragment>}/> as ConditionNode;
		case 'has':
			return <condition title={<fragment>{elementName} properties</fragment>}>
				<table>
					{condition.expression.map(exp => (
						<row>
							<cell>{translateElementProperty(exp.property)}</cell>
							<cell>{translateComparator(exp.type)}</cell>
							<cell>{formatVariables(exp.val + (exp.type === '+-' ? ' Â± ' + exp.deviation : ''), appConfig.configVariables)}</cell>
						</row>
					))}
				</table>
			</condition> as ConditionNode;
		default:
			return assertUnknownElementCondition(condition);
	}
};

const translatePSVideoCondition = (condition: PSVideoHaNoErrorCondition): ConditionNode => {
	const title = <fragment>PlayStation 4 video had no error</fragment>;

	return <condition title={title}>
		<paragraph>{condition.searchStrategy === 'all' ? 'For any source' : 'For current source'}</paragraph>
	</condition> as ConditionNode;
};

const translateJavaScriptExpressionCondition = (
	condition: JavaScriptExpressionCondition,
	appConfig: AppConfiguration
): ConditionNode => {
	const code = replaceVariables(condition.subject.val, appConfig.configVariables);

	return <condition title="JavaScript expression">
		<code-block label="JavaScript expression">{code}</code-block>
		{code !== condition.subject.val
			? <code-block label="With variables">{condition.subject.val}</code-block>
			: undefined
		}
		<table>
			<row>
				<cell>Expression result</cell>
				<cell>{translateComparator(condition.type)}</cell>
				<cell>{formatVariables(condition.val, appConfig.configVariables)}</cell>
			</row>
		</table>
	</condition> as ConditionNode;
};

const translateNetworkInfo = (isRequest: boolean, appConfig: AppConfiguration) => ({
	name,
	compare,
	val,
}: { name: string, compare: Comparator, val: string | number }): Node | Node[] =>
	<row>
		<cell>
			<text>{isRequest ? 'Request' : 'Response'} </text>
			{name.startsWith('@')
				? <text>{name.slice(1)}</text>
				: <fragment>header {formatVariables(name, appConfig.configVariables)}</fragment>
			}
		</cell>
		<cell>{translateComparator(compare)}</cell>
		<cell>{formatVariables(String(val), appConfig.configVariables)}</cell>
	</row>;

/**
 * Put @status and @method on top, then show headers, then - @body.
 * This order is more familiar to user and closer to HTTP raw request/response.
 *
 * Ordering might be off for headers if variables are used for header names, but that is minor
 */
const sortNetworkInfo = (a: NetworkRequestInfo, b: NetworkRequestInfo): number => {
	// Method and status always on top
	if (['@method', '@status'].includes(a.name)) {
		return -1;
	}

	if (['@method', '@status'].includes(b.name)) {
		return 1;
	}

	// Body always last
	if (a.name === '@body') {
		return 1;
	}

	if (b.name === '@body') {
		return -1;
	}

	return a.name > b.name ? -1 : 1;
};

const translateNetworkRequestCondition = (
	condition: NetworkRequestCondition,
	appConfig: AppConfiguration
): ConditionNode => {
	const requestInfo = condition
		.subject
		.requestInfo?.sort(sortNetworkInfo)
		.map(translateNetworkInfo(true, appConfig)) ?? [];
	const responseInfo = condition
		.subject
		.responseInfo?.sort(sortNetworkInfo)
		.map(translateNetworkInfo(false, appConfig)) ?? [];
	const tableRows = requestInfo.concat(responseInfo);

	return <condition title={<text>Network request {condition.type === 'made' ? 'was made' : 'was not made'}</text>}>
		<dictionary>
			<row>
				<cell>{condition.subject.compare === '=' ? 'Exact URL' : 'URL matching'}</cell>
				<cell>{formatVariables(condition.subject.val, appConfig.configVariables)}</cell>
			</row>
			<row>
				<cell>Previously matched</cell>
				<cell>
					<bold>{condition.searchStrategy === 'all' ? 'Yes' : 'No'}</bold>
				</cell>
			</row>
		</dictionary>
		{tableRows.length
			? <table label="Network request properties">
				{tableRows}
			</table>
			: undefined
		}
	</condition> as ConditionNode;
};

const assertUnknownConditionSubject = (subject: never): never => {
	throw new Error(`Unknown condition subject: ${JSON.stringify(subject)}`);
};

/**
 * Translate condition part of the test line into human readable strings
 * "elements" is optional property. BE is going to update results feed to include elements map,
 * but for now it can be omitted
 */
export const translateCondition = (
	condition: Condition,
	appConfig: AppConfiguration,
	elements?: Elements
): ConditionNode => {
	switch (condition.subject.type) {
		case 'element':
		case 'video':
			return translateElementCondition(condition as ElementCondition, appConfig, elements);
		case 'psVideo':
			if (condition.type === 'hadNoError') {
				return translatePSVideoCondition(condition);
			} else {
				return translateElementCondition(condition as ElementPropertiesCondition, appConfig, elements);
			}
		case 'javascript':
			return translateJavaScriptExpressionCondition(condition as JavaScriptExpressionCondition, appConfig);
		case 'location':
			return translateCurrentLocationCondition(condition as CurrentLocationCondition, appConfig);
		case 'cookie':
			return translateCookieCondition(condition as CookieCondition, appConfig);
		case 'network':
			return translateNetworkRequestCondition(condition as NetworkRequestCondition, appConfig);
		case 'application':
			return translateApplicationExitedCondition();
		default:
			return assertUnknownConditionSubject(condition.subject);
	}
};
