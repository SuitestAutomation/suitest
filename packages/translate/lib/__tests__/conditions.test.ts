import {translateCondition} from '../conditions';
import {appConfig, conditions, elements} from './testLines';

describe('conditions translation', () => {
	for (const [name, condition] of Object.entries(conditions)) {
		it(`should translate "${name}" line`, () => {
			expect(translateCondition(condition(), appConfig, elements)).toMatchSnapshot();
		});
	}
});
