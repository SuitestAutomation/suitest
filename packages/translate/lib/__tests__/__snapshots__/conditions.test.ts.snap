// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conditions translation should translate "PS4 video had no error" line 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "For any source",
        },
      ],
      "title": undefined,
      "type": "paragraph",
    },
  ],
  "title": Array [
    Object {
      "type": "text",
      "value": "PlayStation 4 video had no error",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "application has exited" line 1`] = `
Object {
  "children": Array [],
  "title": Array [
    Object {
      "type": "text",
      "value": "Application has exited",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "cookie" line 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "my-cookie",
                },
                Object {
                  "type": "text",
                  "value": " cookie",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "!=",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "expecte value",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
      ],
      "label": undefined,
      "type": "table",
    },
  ],
  "title": Array [
    Object {
      "type": "text",
      "value": "Cookie",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "current location" line 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Current location",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "contains",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "expected value",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
      ],
      "label": undefined,
      "type": "table",
    },
  ],
  "title": Array [
    Object {
      "type": "text",
      "value": "Current location",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "element ... does not exist" line 1`] = `
Object {
  "children": Array [],
  "title": Array [
    Object {
      "type": "bold",
      "value": "Element 1",
    },
    Object {
      "type": "text",
      "value": " does not exist",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "element ... is visible" line 1`] = `
Object {
  "children": Array [],
  "title": Array [
    Object {
      "type": "bold",
      "value": "My element",
    },
    Object {
      "type": "text",
      "value": " is visible",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "element matches JS" line 1`] = `
Object {
  "children": Array [
    Object {
      "label": Array [
        Object {
          "type": "text",
          "value": "JavaScript matcher",
        },
      ],
      "type": "code-block",
      "value": "someJS();",
    },
  ],
  "title": Array [
    Object {
      "type": "code",
      "value": "{\\"css\\":\\".some.class\\",\\"xpath\\":\\"//div\\",\\"index\\":1}",
    },
    Object {
      "type": "text",
      "value": " matches JavaScript",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "element properties" line 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "video position",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": ">=",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "123",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "height",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "=",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "123 ± 3v5",
                },
                Object {
                  "type": "text",
                  "value": " (",
                },
                Object {
                  "type": "code",
                  "value": "123 ± <%var3%>",
                },
                Object {
                  "type": "text",
                  "value": ")",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "text",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "contains",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "expected text",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
      ],
      "label": undefined,
      "type": "table",
    },
  ],
  "title": Array [
    Object {
      "type": "text",
      "value": "Unknown element (",
    },
    Object {
      "type": "code",
      "value": "unkn...n-id",
    },
    Object {
      "type": "text",
      "value": ") properties",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "network request matching URL was not made excluding previously matched" line 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "URL matching",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "/partial/url/123",
                },
                Object {
                  "type": "text",
                  "value": " (",
                },
                Object {
                  "type": "code",
                  "value": "/partial/url/<%var1%>",
                },
                Object {
                  "type": "text",
                  "value": ")",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Previously matched",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "No",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
      ],
      "label": undefined,
      "type": "dictionary",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Request method",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "=",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "GET",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Request header ",
                },
                Object {
                  "type": "bold",
                  "value": "Accept-Type",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "contains",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "text/javascript",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Request body",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "=",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "request val2 body",
                },
                Object {
                  "type": "text",
                  "value": " (",
                },
                Object {
                  "type": "code",
                  "value": "request <%var2%> body",
                },
                Object {
                  "type": "text",
                  "value": ")",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Response status",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": ">",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "123",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Response header ",
                },
                Object {
                  "type": "bold",
                  "value": "Content-123",
                },
                Object {
                  "type": "text",
                  "value": " (",
                },
                Object {
                  "type": "code",
                  "value": "Content-<%var1%>",
                },
                Object {
                  "type": "text",
                  "value": ")",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "ends with",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "val2/javascript",
                },
                Object {
                  "type": "text",
                  "value": " (",
                },
                Object {
                  "type": "code",
                  "value": "<%var2%>/javascript",
                },
                Object {
                  "type": "text",
                  "value": ")",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Response body",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "=",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "response <%non_existing_var%> <%kind-of|var but not%> body",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
      ],
      "label": Array [
        Object {
          "type": "text",
          "value": "Network request properties",
        },
      ],
      "type": "table",
    },
  ],
  "title": Array [
    Object {
      "type": "text",
      "value": "Network request was not made",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "network request to URL was made including matched" line 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Exact URL",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "http://suite.st/123",
                },
                Object {
                  "type": "text",
                  "value": " (",
                },
                Object {
                  "type": "code",
                  "value": "http://suite.st/<%var1%>",
                },
                Object {
                  "type": "text",
                  "value": ")",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Previously matched",
                },
              ],
              "type": "cell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "bold",
                  "value": "Yes",
                },
              ],
              "type": "cell",
            },
          ],
          "type": "row",
        },
      ],
      "label": undefined,
      "type": "dictionary",
    },
  ],
  "title": Array [
    Object {
      "type": "text",
      "value": "Network request was made",
    },
  ],
  "type": "condition",
}
`;

exports[`conditions translation should translate "video exists" line 1`] = `
Object {
  "children": Array [],
  "title": Array [
    Object {
      "type": "bold",
      "value": "video",
    },
    Object {
      "type": "text",
      "value": " exists",
    },
  ],
  "type": "condition",
}
`;
